src/backup.rs

```diff
@@ -0,0 +1,149 @@
+use failure::Error;
+
+use adb_command::AdbCommand;
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct BackupOptions {
+    device_id: Option<String>,
+    applications: String,
+    shared_storage: String,
+    system_apps: String,
+    only_specified_apps: String,
+}
+
+impl BackupOptions {
+    pub fn default() -> Self {
+        BackupOptions {
+            device_id: None,
+            applications: "-noapk".to_string(),
+            shared_storage: "-noshared".to_string(),
+            system_apps: "-nosystem".to_string(),
+            only_specified_apps: "-all".to_string(),
+        }
+    }
+
+    pub fn with_device_id(self, device_id: String) -> Self {
+        BackupOptions {
+            device_id: Some(device_id),
+            ..self
+        }
+    }
+
+    pub fn with_applications(self) -> Self {
+        BackupOptions {
+            applications: "-apk".to_string(),
+            ..self
+        }
+    }
+
+    pub fn with_shared_storage(self) -> Self {
+        BackupOptions {
+            shared_storage: "-shared".to_string(),
+            ..self
+        }
+    }
+
+    pub fn with_system_apps(self) -> Self {
+        BackupOptions {
+            system_apps: "-system".to_string(),
+            ..self
+        }
+    }
+
+    pub fn with_only_specified_apps(self, apps: String) -> Self {
+        BackupOptions {
+            only_specified_apps: apps,
+            ..self
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Backup {}
+
+impl Backup {
+    pub fn backup(backup_options: BackupOptions) -> Result<(), Error> {
+        let command_args = vec![
```

Comment:
it shouldn't be necessary to clone these strings, no? backup_options has a longer lifetime than the command_args

Reply:
Nice hint, thx!
Removed a lot of `to_string` and `clone` calls 👍.